File: FinalReview/lpe_python/lexer.py
'''
CFG
E -> M + E|M - E|M
M -> N * M|N / M|sq M|N
N -> n|(E)
where n is any integer
'''

from functools import reduce
import re

def lex(instr):
  toklst = []
  number_re = re.compile(r"^(-?\d+)")
  terminal_re = re.compile(r"^(sq|[()\-+/*])")
  wspace_re = re.compile(r"^(\s+)")
  pos = 0
  strlen = len(instr)
  while pos < strlen:
    match = re.match(number_re,instr[pos:])
    if match:
      toklst.append(match.group(1))
      pos += len(match.group(1))
    else:
      match = re.match(terminal_re,instr[pos:])
      if match:
        toklst.append(match.group(0))
        pos += len(match.group(0))
      else:
        match = re.match(wspace_re,instr[pos:])
        if match:
          pos += len(match.group(1))
        else:
          raise SyntaxError("Invalid character")
  return toklst

File: FinalReview/lpe_rust/lexer.rs
// Grammar:
//  CFG
//  E -> M + E|M - E|M
//  M -> N * M|N / M|sq M|N
//  N -> n|(E)
//  where n is any integer

use regex::Regex;

fn lex(instr:&str) -> Vec<&str>{
    let mut toklst:Vec<&str> = Vec::new();
    let mut pos = 0;
    let number_re = Regex::new(r"^\d+").unwrap();
    let terminal_re = Regex::new(r"^(sq|[()\-+/*])").unwrap();
    let wspace = Regex::new(r"^\s+").unwrap();

    while pos < instr.len() {
        if let Some(caps) = number_re.captures(&instr[pos..]) {
            if let Some(tok) = caps.get(1) {
                toklst.push(tok.as_str());
                pos += tok.as_str().len();
            }
        } else if let Some(caps) = terminal_re.captures(&instr[pos..]) {
            if let Some(tok) = caps.get(1) {
                toklst.push(tok.as_str());
                pos += tok.as_str().len();
            }
        } else if let Some(caps) = wspace.captures(&instr[pos..]) {
            if let Some(tok) = caps.get(1) {
                pos += tok.as_str().len();
            }
        } else {
            panic!("Invalid token at position {}", pos);
        }
    }
}